File: shell/gnome/src/providerStatus.js
import { httpsUrl, list, oneOf, text, timestamp } from './fields.js';

const INDICATORS = new Set(['none', 'minor', 'major', 'critical', 'maintenance']);
const INDICATOR_TONES = {
    none: 'neutral',
    minor: 'warning',
    maintenance: 'warning',
    major: 'critical',
    critical: 'critical',
};
const STATUS_TONES = new Set(['neutral', 'warning', 'critical']);

function parseEntry(raw) {
    const provider = text(raw.provider);
    const indicator = oneOf(INDICATORS, raw.indicator, null);
    if (provider === null || indicator === null) return null;
    return {
        provider,
        indicator,
        tone: oneOf(STATUS_TONES, raw.tone, INDICATOR_TONES[indicator]),
        title: text(raw.title),
        stage: text(raw.stage),
        startedAt: timestamp(raw.started_at),
        url: httpsUrl(raw.url),
    };
}

export function parseProviderStatus(raw) {
    return list(raw).map(parseEntry).filter(Boolean);
}

export function providerIncident(statuses, provider) {
    return statuses.find(entry => entry.provider === provider && entry.indicator !== 'none') ?? null;
}

File: shell/macos/Sources/HeadroomKit/Models/ProviderStatus.swift
public enum StatusIndicator: String, LenientStringEnum {
    case none, minor, major, critical, maintenance, unknown
    public static let fallback = StatusIndicator.unknown
}

public struct ProviderStatus: Decodable, Sendable, Hashable, Identifiable {
    public let provider: String
    public let indicator: StatusIndicator
    public let tone: Tone
    public let title: String?
    public let stage: String?
    public let startedAt: Timestamp?
    public let url: String

    public var id: String { provider }
    public var isClear: Bool { indicator == .none }

    enum CodingKeys: String, CodingKey {
        case provider, indicator, tone, title, stage, url
        case startedAt = "started_at"
    }
}

public enum PollingMode: String, LenientStringEnum {
    case live, idle
    public static let fallback = PollingMode.idle
}

public enum PollingReason: String, LenientStringEnum {
    case hold, backoff, activity, schedule, unknown
    public static let fallback = PollingReason.unknown
}

public struct AccountRefresh: Decodable, Sendable, Hashable {
    public let mode: PollingMode
    public let intervalSecs: Int64
    public let nextAt: Timestamp?
    public let reason: PollingReason

    enum CodingKeys: String, CodingKey {
        case mode, reason
        case intervalSecs = "interval_secs"
        case nextAt = "next_at"
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        mode = try container.decode(PollingMode.self, forKey: .mode)
        intervalSecs = try container.decode(Int64.self, forKey: .intervalSecs)
        nextAt = try container.decodeIfPresent(Timestamp.self, forKey: .nextAt)
        reason = try container.value(.reason, default: PollingReason.unknown)
    }
}
